CIP-meta-assets (ERC20-like assets)/README.md

```diff
@@ -0,0 +1,578 @@
+        end
+
+        subgraph burned_asset
+            direction LR
+            brnInStart[ ] -.-x brnInStop[ ]
+        
+            style brnInStart fill:#FFFFFF00, stroke:#FFFFFF00;
+            style brnInStop  fill:#FFFFFF00, stroke:#FFFFFF00;
+        end
+
+        style input/output fill:#FFFFFF00, stroke:#FFFFFF00;
+        style ref_input fill:#FFFFFF00, stroke:#FFFFFF00;
+        style minted_asset fill:#FFFFFF00, stroke:#FFFFFF00;
+        style burned_asset fill:#FFFFFF00, stroke:#FFFFFF00;
+    end
+    style utxos fill:#FFFFFF00, stroke:#FFFFFF00;
+
+
+    subgraph transactions
+        direction LR
+
+        subgraph Transaction
+            .[ ]
+            style . fill:#FFFFFF00, stroke:#FFFFFF00;
+        end
+
+        subgraph transaction
+            direction LR
+            start[ ] --> stop[ ]
+        
+            style start fill:#FFFFFF00, stroke:#FFFFFF00;
+            style stop  fill:#FFFFFF00, stroke:#FFFFFF00;
+        end
+        style transaction fill:#FFFFFF00, stroke:#FFFFFF00;
+
+    end
+    style transactions fill:#FFFFFF00, stroke:#FFFFFF00;
+
+    
+
+```
+
+### High level idea
+
+The core idea of the implementation is to emulate the ERC20 standard; where tokens are entries in a map with addresses (or credentials in our case) as key and integers (the balances) as value. ([see the OpenZeppelin implementation for reference](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/9b3710465583284b8c4c5d2245749246bb2e0094/contracts/token/ERC20/ERC20.sol#L16));
+
+Unlike the ERC20 standard; this CIP:
+
+- allows for multiple entries with the same key (same credentials can be used for multiple accounts, though not particularly useful)
+- DOES NOT include an equivalent of the `transferFrom` method; if needed it can be added by a specific implementation but it won't be considered part of the standard.
+- allows for multiple inputs (from the same sender, scripts and multisigs MAY also be used) and multiple outputs (possibly many distinct reveivers)
+
+> **NOTE: Multiple Inputs**
+>
+> the UTxO model allows for multiple transfers in one transaction
+>
+> this would allow for a more powerful model than the account based equivalent but implies higher execution costs
+>
+> with the goal of keeping the standard simple we allow only a single sender
+>
+> we MUST however account for multiple inputs from the same sender due to
+> the creation of new UTxOs in receiving transactions.
+
+### Design
+
+with the introduction of [CIP-69](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0069) in Plutus V3 the number of contracts required are only 2, covering different purposes.
+
+1) a `stateManager` contract
+    - with minting policy used to proof the validity of account's state utxos
+    - spending validator defining the rules to update the states
+2) a `transferManager` parametrized with the `stateManager` hash
+    - having minting policy for the tokens to be locked in the spending validator
```

Comment:
Can this be made more general? Having the `transferManager` also mint the tokens means users will have a separate address for every programmable token. This can quickly become a nightmare for frontends to integrate. It would be better if there was a single `transferManager` that could be used with all programmable tokens.

Programmable tokens don't necessarily need to be minted by the `transferManager`. Instead, they can be deposited into the `transferManager` by the entity minting the token. For example, a user selling NFTs can ensure that all NFTs sold go to a `transferManager` address.

There may also be use cases where a token should be allowed to leave the `transferManager`. For example, lets say there is an art NFT that has been collecting royalties for a while and now the royalty owners would like to release the NFT from the royalty requirement (there may be legal/tax benefits to this in the future). This approach doesn't seem to have a "release" mechanism.

A "release" mechanism could also be useful for upgrading the `transferManager`. If a new version comes out, users can possibly transfer their assets from `transferManager` v1 to `transferManager` v2 (depending on the control utxos). The current design seems to prevent upgrades like this. If this CIP actually does support upgrades, it definitely needs to mention how upgrades are possible. Upgrades don't seem to be mentioned anywhere.